### ISSUE TITLE
Read-only mode is broken for old loom files and should be checked for in more functions

### ISSUE THREAD

**Issue by JobLeonard** [linnarsson-lab/loompy#31]
# Read-only mode and missing HDF5 attributes/groups

After switching to loompy 2, old loom files suddenly refused to open in the loom-viewer. Trying to open them in the REPL caused no problems, and after doing so they could suddenly be opened by the loom-viewer too!

The cause turned out to be that when an older file misses recent additions such as graphs or timestamps, loompy creates a default value for these, [as seen in this section of `GraphManager.__init__`](https://github.com/linnarsson-lab/loompy/blob/9049f9d07e1dd6f50a7694da5337738fed0d768a/loompy/graph_manager.py#[L22-L34]):

```python
a = ["row_edges", "col_edges"][self.axis]
if a in ds._file:
	for key in ds._file[a]:
		self.__dict__["storage"][key] = None
else:
	ds._file.create_group(a)
```

HDF5 (rightfully) complains if you try to do this in read-only mode:
```python
  File "h5py/_objects.pyx", line 54, in h5py._objects.with_phil.wrapper
  File "h5py/_objects.pyx", line 55, in h5py._objects.with_phil.wrapper
  File "h5py/h5g.pyx", line 151, in h5py.h5g.create
ValueError: Unable to create group (no write intent on file)
```

This incidentally also explains why the cache did not work as expected: if `last_modified` is missing, the most recent timestamp is returned, meaning the cache is *always* outdated. As a work-around I am writing an "initiate_last_modified" function that opens a loom file with write-access, iterates over over *all possible time-stamp elements* to ensure they are initialised, then closes, after which the loom-viewer can resume with using read-only mode.

# Adding read-only mode checks to all methods that modify loom files

Currently, only the `last_modified` methods, as well as `add_columns` and `add_loom` check for write access. I suggest we extend this to *all* setters and similar functions that modify the loom file (like `permute`). We cannot prevent people from bypassing the loompy library and modifying the HDF5 file, anyone respecting the library API would at least have better protection this way.

To make this easy to implement, and to ensure that this does not notably slow down regular `r+` usage, we can add a private boolean to `LoomConnection` during `__init__`:

```python
# LoomConnection
self._write_access = mode == 'r+'

# All helper classes that are passed a LoomConnection as `ds`
self._write_access = ds._write_access
```

Then we just have to add these two lines to all functions that modify the loom file:
```python
if not self._write_access:
	raise IOError("Cannot modify loom file when connected in read-only mode")
```

The added benefit of this would be that we raise an error before going all the way down to the h5py or HDF5 library, reducing reliance on *those* implementation of being bug-free.

**Comment by JobLeonard** [comment]
I've started working on this in the `read_only_checks` branch.

Uploaded a `loompy.py` that *should* (requires testing) complain when trying to modify the loom file in read-only mode. I also updated other files (I think I have everything actually) but I'm still double-checking if I missed anything before pushing the commits.

**Comment by slinnarsson** [comment]
Try connecting to a loom file:

```python
with loompy.connect("/Users/sten/build_20171205/L5_All.agg.loom") as ds:
```
```
---------------------------------------------------------------------------
KeyError                                  Traceback (most recent call last)
~/Code/loompy/loompy/layer_manager.py in __getattr__(self, name)
     85                 try:
---> 86                         vals = self.__dict__["storage"][name]
     87                         if vals is None:

KeyError: 'storage'
```

The reason is that you're now preventing setting regular class attributes on read-only files. The code in `__setattr__` is very non-trivial, because to support accessing layers as `ds.layers.exons`, the LayerManager intercepts member access (by implementing `__getattr__`). Yet, it still needs to have some place to store its own attributes (such as `LayerManager.ds` and `LayerManager.storage`). To accomplish this [without creating an infinite loop](https://stackoverflow.com/questions/17020115/how-to-use-setattr-correctly-avoiding-infinite-recursion), `__setattr__` looks for attributes starting with an exclamation mark, such as `!storage` and stores those as object attributes in `self.__dict__`. 

But this code:

```python
		if self._write_access:
			if name.startswith("!"):
				super(LayerManager, self).__setattr__(name[1:], val)
			else:
```

prevents that from happening, so the `__init__` method is not given the chance to set `ds` and `storage`. 

Furthermore, you're setting `self._write_access` *after* calling `setattr`:

```python
	def __init__(self, ds: Any) -> None:  # Note: can't give type for ds because it will be circular and mypy doesn't support it
		setattr(self, "!ds", ds)
		storage: Dict[str, np.ndarray] = {}
		setattr(self, "!storage", storage)

		self._write_access = False
```

which explains the other exception I get, from inside `__setattr__`:

```
AttributeError: '<class 'loompy.layer_manager.LayerManager'>' object has no attribute '_write_access'
```

Probably the same issues apply to the other manager classes.

I don't have time to untangle this at the moment, but please at least make sure you can open a loom file and read all its contents without causing exceptions.

**Comment by JobLeonard** [comment]
Thanks for the detailed feedback, this is very helpful! Don't worry about untangling it - I'll look into that.

**Comment by slinnarsson** [comment]
Fixed in [2.0.9](https://github.com/linnarsson-lab/loompy/releases/tag/2.0.9).

**Comment by JobLeonard** [comment]
Does this invalidate the changes I made in my branch?

**Comment by slinnarsson** [comment]
I thought you merged the branch in d92bac9f04880bffcf00c1655d98d8fdc84a67d1? That merge is included in 2.0.9.
